saleor/graphql/product/tests/mutations/test_product_variant_update.py

```diff
@@ -2533,3 +2533,341 @@ def test_update_product_variant_can_not_turn_off_preorder(
+@patch("saleor.plugins.manager.PluginsManager.product_variant_metadata_updated")
+def test_update_product_variant_with_existing_metadata_and_no_event_when_write_the_same(
+    product_variant_metadata_updated_webhook_mock,
+    product_variant_updated_webhook_mock,
+    staff_api_client,
+    product,
+    permission_manage_products,
+):
+    # Given
+    # - Metadata in variant is already existing
+    # - mutation writes the same metadata key and value
+    query = """
+        mutation updateVariant (
+            $id: ID!,
+            $metadata: [MetadataInput!]
+            ) {
+                productVariantUpdate(
+                    id: $id,
+                    input: {
+                        metadata: $metadata,
+                        privateMetadata: $metadata
+                    }) {
+                    productVariant {
+                        id
+                        name
+                    }
+                }
+            }
+    """
+    variant = product.variants.first()
+    variant_id = graphene.Node.to_global_id("ProductVariant", variant.pk)
+
+    metadata_key = "mk"
+    metadata_value = "mv"
+
+    variant.name = "Name"
+
+    variant.metadata = {metadata_key: metadata_value}
+
+    variant.private_metadata = {metadata_key: metadata_value}
+
+    variant.save()
+
+    assert variant.metadata == {metadata_key: metadata_value}
+
+    assert variant.private_metadata == {metadata_key: metadata_value}
+
+    # When
+    # - Metadata is updated with the same values
+    variables = {
+        "id": variant_id,
+        "metadata": [{"key": metadata_key, "value": metadata_value}],
+        "privateMetadata": [{"key": metadata_key, "value": metadata_value}],
+    }
+
+    response = staff_api_client.post_graphql(
+        query, variables, permissions=[permission_manage_products]
+    )
+    variant.refresh_from_db()
+    content = get_graphql_content(response)
+    data = content["data"]["productVariantUpdate"]["productVariant"]
+
+    assert data["id"] == variant_id
+
+    # Then
+    # - product_variant_updated should not run - nothing changed
+    # - product_variant_metadata_updated should not run - metadata not changed, values the same,
+    # no empty event emitted
+    product_variant_updated_webhook_mock.assert_not_called()
+    product_variant_metadata_updated_webhook_mock.assert_not_called()
+
+
+@patch("saleor.plugins.manager.PluginsManager.product_variant_updated")
+@patch("saleor.plugins.manager.PluginsManager.product_variant_metadata_updated")
+def test_update_product_variant_with_existing_metadata_and_event_when_write_different_value(
+    product_variant_metadata_updated_webhook_mock,
+    product_variant_updated_webhook_mock,
+    staff_api_client,
+    product,
+    permission_manage_products,
+):
+    # Given
+    # - Metadata in variant is already existing
+    # - mutation writes the same metadata key but different value
+    query = """
```

Comment:
You are using the same query multiple times, you can create a constant from this and reuse it.
Instead of providing `metadata` in mutation argument use `input` instead, which allow to cover all test cases.
```
     mutation updateVariant (
            $id: ID!,
            $input: ProductVariantInput!
            ) {
                productVariantUpdate(
                    id: $id,
                    input: $ProductVariantInput {
                    productVariant {
                        id
                        name
                    }
                }
            }
```